### ISSUE TITLE
[Feature Suggestion] Queue System

### ISSUE THREAD

**Issue by oga-sven** [rsxdalv/tts-generation-webui#44]
I have no experience with gradio otherwise i would even attempt it myself. I plan to even learn from it haha
My suggestion is to add a queue system to the voice generation. Instead of clicking generate, you would basically click add to queue. Process queue would then take care of it. This would help with line splitting as well. In this specific case i am talkign about tortoise. As i am getting way more stable results from that one

https://www.reddit.com/r/StableDiffusion/comments/zq0wl9/i_made_a_queue_system_for_automatic1111s_stable/
https://github.com/Kryptortio/SDAtom-WebUi-us 
This is an example on how i envision it

**Comment by rsxdalv** [comment]
Thanks for letting me know about tortoise, I assumed it was unused. As for
the queue - before that, let's speed up tortoise at least a little. I will
check if I can cache the model.

Meanwhile, there is a way to queue within gradio, although I'm not sure it
will work as well (since it might try to do it in parallel, resulting in a
mess): open multiple tabs, and generate in each tab.

There's a queue size (in the code, I'll expose the setting), if I set it to
1, then everything will be run in a single queue.

I'll check if the links can help improve this.

**Comment by oga-sven** [comment]
Personally i dont care much about the speed. If i put a 100 entry queue in, its gonna run in the background anyway so im good with that. I always run on highest quality and CVVP amount maxed (I have no clue what it is. but it makes it sound way better than 0 or 0.5). The links is just to visually understand it. i believe the exention itself is just a JS code that interacts with the webclient. (Which i would also be fine with ofcourse haha). What i would like to add though is maybe some adjustment on the filename/output if the queue is used. just for organisation

**Comment by rsxdalv** [comment]
Wait just a question since you mentioned it before - what's the use case?
Are you just looking to get a big chunk of text spoken?
In that case a "split by lines" would be a good match and follow the Bark
UI.

**Comment by oga-sven** [comment]
If you could implement split by lines in tortoise that would also be nice! however, I wanted to have multiple generations by line. (This is not a hard requirement for me though). 
So line one would be generated 3 times
line two would be generated 3 times.. etc
I dont believe thats possible now with the split by lines functionality. Also adjusting parameters per line:

line 1 - param 1 3 times
line 1 - param 2 3 times
line 2 - param 1 3 times
etc..

Again, this is not a hard requirement but a nice to have. i would definitely already be satisfied by a split by lines haha

**Comment by oga-sven** [comment]
thats nice! ill gladly mess around a bit more with that. in that case the split by line for tortoise would be good 💪

**Comment by JonathanFly** [comment]
> 

I'd like to come up with a good queue system. For me, it's not about splitting text, it's about setting 50 different very specific UI widgets in the way that I tend to do, super tweaked settings, and then the hard part, take your output files and using them as new inputs, in different ways. The last part is what flummoxed me in Gradio. I could still do it by just ignoring the Gradio UI but in my head I'm drawing input output lines between widgets.

Anyway, if you want to put our heads together on a queue, I'm all for it. My current best working concept is to take the code that saves and loads user setting in the most recent (last month update) Automatic1111. They use this code to save defaults, but it's an example of scanning all possible Gradio UI widgets and either reading or changing them. So my idea is basically, hey, that's a general purpose queue. You scan every Gradio widget, archive it, then on demand load it later. However the files are still a problem. That code doesn't deal with it. I don't want to encoded huge audio files into a json. So I guess I have to make copies or something.

**Comment by rsxdalv** [comment]
> > 
> 
> I'd like to come up with a good queue system. For me, it's not about splitting text, it's about setting 50 different very specific UI widgets in the way that I tend to do, super tweaked settings, and then the hard part, take your output files and using them as new inputs, in different ways. It's lots of 3 or 4 step recursive loops. I do these in Jupyter notebooks now and can't figure out how to make it to work in Gradio. The files part is what really flummoxed me in Gradio. I could still do it by just ignoring the Gradio UI but in my head I'm drawing input output lines between widgets.
> 
> Anyway, if you want to put our heads together on a queue, I'm all for it. My current best working concept is to take the code that saves and loads user setting in the most recent (last month update) Automatic1111. They use this code to save defaults, but it's an example of scanning all possible Gradio UI widgets and either reading or changing them. So my idea is basically, hey, that's a general purpose queue. You scan every Gradio widget, archive it, then on demand load it later. However the files are still a problem. That code doesn't deal with it. I don't want to encoded huge audio files into a json. So I guess I have to make copies or something.
> 
> Alternatively maybe that Automatic is good.

I would just stick to having a gradio function that sends the inputs to an Rx (or any) stream, where the "prediction" function subscribes to this stream.

It looks like this:


```python

$subject = new Subject()

gr.Button("queued").click(fn=lambda **args: $subject.on_next(args))

$subject.subscribe(fn=predict)
```

Although the APIs are probably a bit different from this pseudo code.

https://rxpy.readthedocs.io/en/latest/reference_subject.html#reactivex.subject.Subject.on_next

If they figured out how to serialize gradio that's great, I've also worked on it, although I think I'll start ditching gradio soon enough, it's really a backend framework with a frontend representation.

**Comment by oga-sven** [comment]
Hey man so whats the status/ETA on this? Is there anything i can help with? It *looks* like you got it figured out haha

**Comment by oga-sven** [comment]
https://github.com/rsxdalv/tts-generation-webui/commit/6b29cb652e894c2cae26ee8cbcf5b494e8d77378 is this the commit? I have only been looking at the release page. But im willing to pull master and try it out

**Comment by rsxdalv** [comment]
Yes, if you have the one-click installer than "update" otherwise git pull and pip install
